tests/projection.rs

```diff
@@ -0,0 +1,287 @@
+                            ?ty_2,
+                        ],
+                        coherence_mode: false,
+                    },
+                    known_true: true,
+                    substitution: {
+                        ?ty_2 => (rigid (adt Vec) (alias (Iterator :: Item) !ty_1)),
+                        ?ty_3 => (alias (Iterator :: Item) !ty_1),
+                    },
+                },
+            },
+        )
+    "#]]
+    .assert_debug_eq(&test_where_clause(
+        NORMALIZE_BASIC,
+        "forall<ty T> exists<ty U> { Iterator(T) } => { <T as Iterator<>>::Item<> = <U as Iterator<>>::Item<> }",
+    ));
+}
+
+const NORMALIZE_INTO_ITERATOR: &str = "[
+    crate test {
+        trait IntoIterator<> where [] {
+            type Item<> : [] where [];
+        }
+
+        trait Iterator<> where [] {
+            type Item<> : [] where [];
+        }
+
+        struct Vec<ty T> where [] {}
+
+        struct Foo<> where [] {}
+
+        impl<ty T> IntoIterator<> for Vec<T> where [] {
+            type Item<> = T where [];
+        }
+
+        impl<ty T> IntoIterator<> for T where [ T: Iterator<> ] {
+            type Item<> = <T as Iterator>::Item<> where [];
+        }
+    }
+]";
+
+#[test]
+fn normalize_into_iterator() {
+    expect_test::expect![[r#"
+        Ok(
+            {
+                Constraints {
+                    env: Env {
+                        variables: [
+                            !ty_1,
+                            ?ty_2,
+                        ],
+                        coherence_mode: false,
+                    },
+                    known_true: true,
+                    substitution: {
+                        ?ty_2 => (alias (IntoIterator :: Item) (rigid (adt Vec) !ty_1)),
+                    },
+                },
+                Constraints {
+                    env: Env {
+                        variables: [
+                            !ty_1,
+                            ?ty_2,
+                        ],
+                        coherence_mode: false,
+                    },
+                    known_true: true,
+                    substitution: {
+                        ?ty_2 => !ty_1,
+                    },
+                },
+            },
+        )
+    "#]]
+    .assert_debug_eq(&test_where_clause(
+        NORMALIZE_INTO_ITERATOR,
+        "forall<ty T> exists<ty U> {} => { <Vec<T> as IntoIterator>::Item<> = U }",
+    ));
+}
+
+const PROJECTION_EQUALITY: &str = "[
+    crate test {
+        trait Trait1<> where [] {
+            type Type<> : [] where [];
+        }
+        trait Trait2<ty T> where [] {}
+        impl<ty T, ty U> Trait2<T> for U where [ U: Trait1<>, (alias (Trait1::Type) S) => T ] {}
+        struct S<> where [] {}
+        impl<> Trait1<> for S<> where [] {
+            type Type<> = u32 where [];
+        }
+    }
+]";
+
+
+#[test]
+fn projection_equality() {
+    expect_test::expect![[r#"
+        Ok(
+            {},
```

Comment:
This isn't right :) Should be ?0 = u32. The first goal (the AliasEq) finds this constraint, but after solving the `S: Trait1<>` goal we don't have any

Reply:
This is an outlives predicate